File: BindRobot.py
class Solution:
    def isRobotBounded(self, instructions: str) -> bool:
        if instructions ==None or len(instructions)==0:
            return True
        dirs=[(0,1),(-1,0),(0,-1),(1,0)]
        x=0
        y=0
        i=0
        for k in range(len(instructions)):
            c=instructions[k]
            if c=='L':
                i=(i+1)%4
            elif c=='R':
                i=(i+3)%4
            else:
                x+=dirs[i][0]
                y+=dirs[i][1]
        #if it comes back to origin or faces all directions except north
        #after first set of intructions
        if (x==0 and y==0)  or i!=0:
            return True
        return False
        
#time Complexity is O(n)
#Space is O(1)

File: bind_robot.java
//TC: O(n)
//SC: O(1)

class Solution {
    public boolean isRobotBounded(String instructions) {
       int x=0, y=0, direction = 0;

//          N 0
//     W 3       E 1
//          S 2
    
        
        for(char ch : instructions.toCharArray()){
            if(ch == 'G'){         
                switch (direction){
                case 0: y++; break;
                case 1: x++; break;
                case 2: y--; break;
                case 3: x--; break;
                }
               
            }else if(ch == 'L'){
                direction = direction == 0 ? 3 : (direction-1);
                
            }else if(ch == 'R'){
                direction = (direction + 1)%4;
            }
        }
        if(direction!= 0 || (x == 0 && y == 0)){
            return true;
        }
        return false;
    }
}